### PR TITLE
feat:  Stop too far from trip shape

### LINKED ISSUE #521: Validate Stop not too far from trip shape
**Describe the solution you'd like**
Validate stops are not too far from their trip shape

**Additional context**
`v1.4.0` implementation:
- https://github.com/MobilityData/gtfs-validator/blob/v1.4.0/usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateStopTooFarFromTripShape.java
- https://github.com/MobilityData/gtfs-validator/blob/v1.4.0/usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateStopTooFarFromTripShapeNoticeTest.java

### PR DESCRIPTION
**Pull request by lionel-nj**
closes #521 

**Summary:**

> This PR implements the rule to check when a stop is too far from the trip shape.
> 
> [GTFS Best Practices](https://gtfs.org/best-practices/#shapestxt) say:
> 
> >Route alignments (in shapes.txt) should be within 100 meters of stop locations which a trip serves.
> 
> Data used in unit tests is available as GeoJSON on GitHub here (GitHub provides a toggle between text and map view):
> https://gist.github.com/barbeau/d9c0b90a26a3e2ba105cae5f0e8aec4a
> 
> image
> 
> If desired, you can visualize the buffer by exporting a JTS-version of the buffer in WKT format using code at https://gist.github.com/barbeau/d9c0b90a26a3e2ba105cae5f0e8aec4a#gistcomment-3425554 and then visualizing the WKT using https://arthur-e.github.io/Wicket/sandbox-gmaps3.html.
> 
> image
> 
> The spatial4j version of the buffer can't easily be visualized using GeoJSON or WKT because it uses a LineString and a proprietary "buffer" extension to GeoJSON and WKT, which most tools don't support.
> 

from #340 

New notice: `StopTooFarFromTripShapeNotice`
New util class: `GeospatialUtil`

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: -new feature short description-" (PR title must follow the [conventional commit specification](https://www.conventionalcommits.org/en/v1.0.0/))
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java
Comment (maximearmstrong) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java:131]: Should not the stops be of `location_type` 0 or 4?

File: main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java
Comment (maximearmstrong) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java:278]: Same as above regarding `location_type`

File: main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java
Comment (maximearmstrong) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java:423]: Same as above

**Review by maximearmstrong — CHANGES_REQUESTED** [review]

File: main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java
Comment (lionel-nj) [reply main/src/test/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtilTest.java:131]: Indeed, thanks!

**Review by maximearmstrong — APPROVED** [review]

**Comment by maximearmstrong** [comment]
LGTM

**Comment by aababilov** [comment]
Thank you for that great contribution, Lionel!

I see that you added a dependency on spatial4j which is not available on Google, so I cannot compile the fresh validator. I will evaluate the alternatives.

Please may I ask you to let me know when you are adding a new dependency before you merge you commits? The thing is that when I cannot compile Validator at Google, that is a blocker for me.

**Comment by aababilov** [comment]
We had a similar issue with https://github.com/MobilityData/gtfs-validator/pull/541 that added org.apache.httpcomponents.client5 which is banned at Google.

**Comment by lionel-nj** [comment]
Oh, apologies @aababilov. 

> Please may I ask you to let me know when you are adding a new dependency before you merge you commits? The thing is that when I cannot compile Validator at Google, that is a blocker for me.

Definitely yes! We will proceed this way in the future if that is okay with @barbeau and @carlfredl.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java
Comment (f8full) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java]: If I may suggest so, there's some soft convention to put units at the end. It's definitively something that just requires consistency across the whole code base.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java
Comment (aababilov) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: Here you are checking every trip as many times as you have it in stop_times.txt. That is a bug in the algorithm and a serious performance issue.

What was the size of the largest feed that you used for testing? I used a medium-size feed with 1 million stop_times.txt and this code runs for several minutes on it. If we add a large enough feed to the test set, we can catch such bugs earlier.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java
Comment (aababilov) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:76]: What is the rationale of placing all notices to a single container instead of emitting them inside the lambda function?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java
Comment (aababilov) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java:42]: You are dividing an integer by integer, so the result is [rounded](https://stackoverflow.com/questions/37795248/integer-division-in-java) and it is 0 (zero). You probably wanted to write `1.0 / 1000` instead.

When I executed your code on a real feed, it produced 8521 errors for 68157 trips, so that looked suspicious.

What was the size and diversity of the feeds that you use? Do you mind to cover that bug with unit tests?

Also, why are you using `float` instead of `double` here?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: Another way to find this bug earlier was to see that you are producing the same notice multiple times. Can you add unit tests for that?

Comment (f8full) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: > If we add a large enough feed to the test set, we can catch such bugs earlier.

I think that might explains what happens with Buenos Aires
image
image
https://github.com/MobilityData/gtfs-validator/runs/1881156285?check_suite_focus=true

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java
Comment (lionel-nj) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:76]: It is true that this could be changed to emit notices in the lambda function. Thanks for flagging that @aababilov.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java
Comment (lionel-nj) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: This piece of code origins from: https://github.com/MobilityData/gtfs-validator/pull/340. 
@barbeau could you mind taking a look at this please?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java
Comment (lionel-nj) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java:42]: @aababilov we originally tested the validator against 7 datasets. The biggest dataset in terms of `stop_times.txt` is from STM (https://openmobilitydata.org/p/societe-de-transport-de-montreal/39/latest/download) and contains +/- 5.6 millions rows in `stop_times.txt`.

Comment (lionel-nj) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java:42]: The datasets the validator was tested against are listed in the `end_to_end` workflow : https://github.com/MobilityData/gtfs-validator/blob/v1.4.0/.github/workflows/end_to_end.yml

Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java:42]: Unfortunately, this feed for Montreal was not testing StopTooFarFromTripShapeValidator - its code was never called. The thing is that the Montreal feed has `field_parsing_error` for stop_url field (it uses non-ASCII characters), so parsing failed, the validation process was stopped and StopTooFarFromTripShapeValidator was never invoked.

You can check the generated report.json to find the parsing errors.

```JSON
        {
            "code": "field_parsing_error",
            "notices": [
                {
                    "csvRowNumber": 163,
                    "fieldName": "stop_url",
                    "fieldType": "URL",
                    "fieldValue": "http://www.stm.info/fr/infos/reseaux/metro/square-victoria\u2013oaci",
                    "filename": "stops.txt"
                },
                {
                    "csvRowNumber": 232,
                    "fieldName": "stop_url",
                    "fieldType": "URL",
                    "fieldValue": "http://www.stm.info/fr/infos/reseaux/metro/longueuil\u2013universite-de-sherbrooke",
                    "filename": "stops.txt"
                }
            ],
            "severity": "ERROR",
            "totalNotices": 2
        },
```

Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java:42]: I wrote https://github.com/MobilityData/gtfs-validator/pull/742 to make clear that parsing failed.

Comment (f8full) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java:42]: In v1.4, the workflows would surface those kind of situation in the GitHub interface by outputting an error, which would in turn lead to all subsequent feeds to be skipped. I wonder if such a process could be replicated.

Comment (barbeau) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/util/GeospatialUtil.java:42]: >You are dividing an integer by integer, so the result is rounded and it is 0 (zero). You probably wanted to write 1.0 / 1000 instead.
>Also, why are you using float instead of double here?

Good catch - originally this used a `double` and divided by `1000.0d` as implemented in https://github.com/MobilityData/gtfs-validator/pull/340/files#diff-fd338ec09e9005136186658756b593cc6e7e9917476a0b92a59f27071a1c3314R45. So it should be changed back.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java
Comment (barbeau) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: I need to look at this a little closer, but I think this is a translation issue from the original implementation due to the differences in data structures between v1 and v2. 

In the original implementation of v1, the in-memory GTFS data repository held a Map data structure where each trip_id appeared once as the key and the value was a SortedMap of StopTimes for that trip_id. This rule looped through each of those entries and therefore only processed each trip once.

But the similarly named variable in v2 actually contains the entire contents of stop_times.txt. So it looks like this implementation is fetching all StopTimes and then getting the trip_id for each StopTime and then getting all the StopTimes for that trip_id, which blows up the processing time significantly.

I'll take a closer look tomorrow.

Comment (barbeau) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: @aababilov Are you working on re-implementing this with a separate spatial library? Or should MobilityData work on fixing the implementation issues with this rule using Spatial4j?

Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: > This piece of code origins from: #340.

It actually does not origin from the Sean's commit. Sean wrote another logic that iterated on each trip only once. The new code, however, iterates on each pair of <trip_id, GtfsStopTime>. IDEA can spell out the types for you:

```
    stopTimeTable
        .byTripIdMap()
        .forEach(
            (String tripId, GtfsStopTime tripStopTimes) -> {
                ....
```

Also, the new code does not use the `tripStopTimes` variable - instead, it calls `stopTimeTable.byTripId(tripId)` that really returns `List<GtfsStopTime>`. If you use your `tripStopTimes` in that place, that won't compile and you would catch this bug.

Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: Overall, I think I can suggest you to spell out the types and also prefer plain cycles to lambdas. That will help to avoid such kind of bugs in the future.

Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: > @aababilov Are you working on re-implementing this with a separate spatial library? Or should MobilityData work on fixing the implementation issues with this rule using Spatial4j?

This validation rule is not the top priority for me right now. What I did was to temporarily remove that code at Google.

I think that we should put as a priority that our MobilityData upstream code works as expected for a large set of feeds (such as several thousands that we have at Google). I do not think that we need to rush implementing as many validations as possible without proper testing since that puts too much technical debt for the future maintainers.

I am always testing my code on all Google feeds and I check the results, so that gives a certain confidence. I see that MobilityData is not actually testing the code on a considerable feed set. Even more, the test results may be misinterpreted (as it happened for the Montreal feed), so that reduces positive impact of such testing.

@barbeau @lionel-nj Are you able to setup a reliable testing procedure on your side and run the Validator on a large set of feeds (hundreds or better thousands)? Without such testing, we always risk to accumulate bugs in the upstream code. May we hold on all code contributions from MobilityData before such testing procedure is established?

Comment (lionel-nj) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: @aababilov could you provide the url of the dataset you were looking at please?

Comment (lionel-nj) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java:59]: > I used a medium-size feed with 1 million stop_times.txt and this code runs for several minutes on it.


@aababilov could you provide the url of the dataset you were looking at please?

**Pull request merged**
